--- /tmp/orthologr-orig/R/pairwise_aln.R        2026-01-10 00:44:17
+++ /tmp/orthologr-new/R/pairwise_aln.R 2026-01-10 00:44:17
@@ -85,7 +85,7 @@
                         )
         }
         
-        # Needleman-Wunsch using Biostrings::pairwiseAlignment( type=global)
+        # Needleman-Wunsch using pwalign::pairwiseAlignment( type=global)
         if (tool == "NW") {
                 #read file
                 if (is.element(seq_type, c("dna", "cds"))) {
@@ -122,7 +122,7 @@
                 
                 # align
                 aln  <-
-                        Biostrings::pairwiseAlignment(pattern = seqs[[1]],
+                        pwalign::pairwiseAlignment(pattern = seqs[[1]],
                                                       subject = seqs[[2]],
                                                       type    = "global")
                 
@@ -130,17 +130,17 @@
                 # pattern(aln) == get aligned sequence that was set as pattern input
                 # subject(aln) == get aligned sequence that was set as subject input
                 seqinr::write.fasta(
-                        sequences = list(Biostrings::pattern(aln), IRanges::subject(aln)),
+                        sequences = list(pwalign::pattern(aln), pwalign::subject(aln)),
                         names     = names,
                         file.out  = file.out
                 )
                 
                 if (store_locally) {
                         local_store_path <- file.path("orthologr_alignment_files", "_pairwise_alignment_with_score")
-                        Biostrings::writePairwiseAlignments(aln, file.path(local_store_path, paste0(pairwise_aln_name, "_", tool, "_", seqtype, ".aln")))
+                        pwalign::writePairwiseAlignments(aln, file.path(local_store_path, paste0(pairwise_aln_name, "_", tool, "_", seqtype, ".aln")))
                 } else {
                         local_store_path <- file.path(tempdir(), "_pairwise_alignment_with_score")
-                        Biostrings::writePairwiseAlignments(aln, file.path(local_store_path, paste0(pairwise_aln_name, "_", tool, "_", seqtype, ".aln")))
+                        pwalign::writePairwiseAlignments(aln, file.path(local_store_path, paste0(pairwise_aln_name, "_", tool, "_", seqtype, ".aln")))
                 }
                 
                 if (!quiet) {
